src/server/journal/serializer.cc

```diff
@@ -18,154 +18,182 @@ using namespace std;
+JournalWriter::JournalWriter(std::optional<DbIndex> dbid) : cur_dbid_{dbid} {
 }
 
-error_code JournalWriter::Write(uint64_t v) {
+std::error_code JournalWriter::Flush(io::Sink* sink) {
+  if (auto ec = sink->Write(buf_.InputBuffer()); ec)
+    return ec;
+  buf_.Clear();
+  return {};
+}
+
+io::Bytes JournalWriter::Accumulated() {
+  return buf_.InputBuffer();
+}
+
+void JournalWriter::Write(uint64_t v) {
   uint8_t buf[10];
   unsigned len = WritePackedUInt(v, buf);
-  return sink_->Write(io::Bytes{buf, len});
+  buf_.EnsureCapacity(10);
+  memcpy(buf_.AppendBuffer().data(), buf, len);
+  buf_.CommitWrite(len);
 }
 
-error_code JournalWriter::Write(std::string_view sv) {
-  RETURN_ON_ERR(Write(sv.size()));
-  return sink_->Write(io::Buffer(sv));
+void JournalWriter::Write(std::string_view sv) {
+  Write(sv.size());
+  buf_.EnsureCapacity(sv.size());
+  memcpy(buf_.AppendBuffer().data(), sv.data(), sv.size());
+  buf_.CommitWrite(sv.size());
 }
 
-error_code JournalWriter::Write(CmdArgList args) {
-  RETURN_ON_ERR(Write(args.size()));
+void JournalWriter::Write(CmdArgList args) {
+  Write(args.size());
   for (auto v : args)
-    RETURN_ON_ERR(Write(facade::ToSV(v)));
-
-  return std::error_code{};
+    Write(facade::ToSV(v));
 }
 
-error_code JournalWriter::Write(std::pair<std::string_view, ArgSlice> args) {
+void JournalWriter::Write(std::pair<std::string_view, ArgSlice> args) {
   auto [cmd, tail_args] = args;
 
-  RETURN_ON_ERR(Write(1 + tail_args.size()));
-  RETURN_ON_ERR(Write(cmd));
+  Write(1 + tail_args.size());
+  Write(cmd);
   for (auto v : tail_args)
-    RETURN_ON_ERR(Write(v));
-
-  return std::error_code{};
+    Write(v);
 }
 
-error_code JournalWriter::Write(std::monostate) {
-  return std::error_code{};
+void JournalWriter::Write(std::monostate) {
 }
 
-error_code JournalWriter::Write(const journal::Entry& entry) {
+void JournalWriter::Write(const journal::Entry& entry) {
   // Check if entry has a new db index and we need to emit a SELECT entry.
   if (entry.opcode != journal::Op::SELECT && (!cur_dbid_ || entry.dbid != *cur_dbid_)) {
-    RETURN_ON_ERR(Write(journal::Entry{journal::Op::SELECT, entry.dbid}));
+    Write(journal::Entry{journal::Op::SELECT, entry.dbid});
     cur_dbid_ = entry.dbid;
   }
 
-  RETURN_ON_ERR(Write(uint8_t(entry.opcode)));
+  Write(uint8_t(entry.opcode));
 
   switch (entry.opcode) {
     case journal::Op::SELECT:
       return Write(entry.dbid);
     case journal::Op::COMMAND:
-      RETURN_ON_ERR(Write(entry.txid));
-      RETURN_ON_ERR(Write(entry.shard_cnt));
+      Write(entry.txid);
+      Write(entry.shard_cnt);
       return std::visit([this](const auto& payload) { return Write(payload); }, entry.payload);
     default:
       break;
   };
-  return std::error_code{};
 }
 
-JournalReader::JournalReader(DbIndex dbid) : buf_{}, dbid_{dbid} {
+JournalReader::JournalReader(io::Source* source, DbIndex dbid)
+    : str_buf_{}, source_{source}, buf_{}, dbid_{dbid} {
```

Comment:
reserve space for buf in init something like 4K, so when you call ensure read it will read from socket more data